test/testutils/controller.go

```diff
@@ -168,12 +175,83 @@ func CreateThenCleanup(ctx context.Context, t *testing.T, c client.Client, obj c
 	t.Cleanup(DeleteObjectAndConfirm(ctx, t, c, obj))
 }
 
-func makeUnstructured(gvk schema.GroupVersionKind, namespace, name string) *unstructured.Unstructured {
-	u := &unstructured.Unstructured{
-		Object: make(map[string]interface{}),
+func SetupDataClient(t *testing.T) *constraintclient.Client {
+	driver, err := rego.New(rego.Tracing(false))
+	if err != nil {
+		t.Fatalf("setting up Driver: %v", err)
+	}
+
+	client, err := constraintclient.NewClient(constraintclient.Targets(&target.K8sValidationTarget{}), constraintclient.Driver(driver))
+	if err != nil {
+		t.Fatalf("setting up constraint framework client: %v", err)
+	}
+	return client
+}
+
+// SetupTestReconcile returns a reconcile.Reconcile implementation that delegates to inner and
+// writes the request to requests after Reconcile is finished.
+func SetupTestReconcile(inner reconcile.Reconciler) (reconcile.Reconciler, *sync.Map) {
+	var requests sync.Map
+	fn := reconcile.Func(func(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+		result, err := inner.Reconcile(ctx, req)
+		requests.Store(req, struct{}{})
+		return result, err
+	})
+	return fn, &requests
+}
+
+// ConfigFor returns a config resource that watches the requested set of resources.
+func ConfigFor(kinds []schema.GroupVersionKind) *configv1alpha1.Config {
```

Comment:
The following functions would be better homed in `gatekeeper/pkg/fakes`

Reply:
Why not `testutils`?

Reply:
If there's a better word then that works for me, the point is that this code is re-usable infrastructure that we want to make available to the entire codebase.  As long as that happens, I care less about where exactly they go.

I get what you mean about fakes/stubs/spies.  IMO those concepts don't get to own the world "fake" as it's so general.  `fakes.Pod()` seems to pretty clearly make a fake pod ([link](https://github.com/open-policy-agent/gatekeeper/blob/master/pkg/fakes/pod.go)).

Even in this PR there are places where you are creating fake versions of objects that could be simplified by leveraging this code or something very similar.

Reply:
still this